### PR TITLE
Make connection cache to support specified client endpoint

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem
ConnectionCache is being used for managing connections for sending messages using quic. In the current implementation the connection's endpoint is created in a lazy initialized fashion and set to one per connection pool. In repair we need all connections to use the same Endpoint so that the server can send back the response to the same Endpoint. 

#### Summary of Changes
This is the 2nd installment to support quic in repair workflow.

Make connection to take an optional specified Quic Endpoint.
Unit tests to verify the behavior.

Fixes #

### CODE REVIEW

**Comment by CriesofCarrots** [comment]
Can you ping me if/when these changes are finalized? I want to make sure we're emulating any new behavior in TpuConnectionCache. Thanks!

File: client/src/connection_cache.rs [hunk 1]
```diff
@@ -78,6 +83,18 @@ impl ConnectionCache {
         }
     }
 
+    /// Create a connection cache with a specific quic client endpoint.
+    pub fn new_with_endpoint(connection_pool_size: usize, client_endpoint: Endpoint) -> Self {
+        // The minimum pool size is 1.
+        let connection_pool_size = 1.max(connection_pool_size);
```
Comment (CriesofCarrots) [inline client/src/connection_cache.rs:89]: Could we make this duplicated code less brittle by adding a non-pub method that both `ConnectionCache::new()` and `ConnectionCache::new_with_endpoint()` call, like:

```
pub fn new(connection_pool_size: usize) -> Self {
    Self::_new_with_endpoint(connection_pool_size, None)
}

pub fn new_with_endpoint(connection_pool_size: usize, client_endpoint: Endpoint) -> Self {
    Self::_new_with_endpoint(connection_pool_size, Some(client_endpoint)
}

fn _new_with_endpoint(connection_pool_size: usize, client_endpoint: Option<Endpoint>) -> Self {
    // The minimum pool size is 1.
    let connection_pool_size = 1.max(connection_pool_size);
    Self {
        use_quic: true,
        connection_pool_size,
        client_endpoint,
        ..Self::default()
    }
}
```

**Review by CriesofCarrots — COMMENTED** [review]
Thanks for the tag. Looks like this will just be adding a new field to `QuicConfig` for the refactored TpuConnectionCache; pretty straightforward.

File: client/src/connection_cache.rs (hunk 1 above)
Comment (lijunwangs) [reply client/src/connection_cache.rs:99]: Done

**Review by CriesofCarrots — APPROVED** [review]
Works for me

**Pull request merged**